keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -666,7 +773,1310 @@ ways.  One is that we are dispatching requests to be served rather
+concurrency they demand.  That is, `i` for which `mu(i,t) < rho(i,t)`.
+These queues all get the same allocation, `mu_fair(t)`.  BTW, we could
+alternatively distinguish the categories by comparing `mu` to
+`mu_fair` rather than `rho`; this will make a difference approximately
+never (i.e., only when some queue's demand exactly equals the fair
+allocation), and there is no significant consequence to that
+difference.  The fact that the allocations are equal is the key to an
+efficient representation for this category.  We use the same `R(t)`
+meter as the original Fair Queuing algorithm.
+
+```
+R(t) = Integral[from tau=epoch to tau=t] mu_fair(tau) dtau
+```
+
+For a given queue `i`, suppose
+- the equation `mu(i,t) = mu_fair(t)` has been true since `t = t_i0`,
+  and
+- `j_next(i)` has been constant throughout that time.
+
+Assume those will continue to hold and estimate as follows.
+
+```
+Integral[t_i0 <= tau <= nextEE(i)] rate(i,tau) dtau
+        = PEnd(i,j_next(i)) - P(i,t_i0)
+
+rate(i,tau) = mu_fair(tau) / width(i,j_next(i))
+
+Integral[t_i0 <= tau <= nextEE(i)] mu_fair(tau) dtau / width(i,j_next(i))
+        = PEnd(i,j_next(i)) - P(i,t_i0)
+
+Integral[t_i0 <= tau <= nextEE(i)] mu_fair(tau) dtau
+        = width(i,j_next(i)) * (PEnd(i,j_next(i)) - P(i,t_i0))
+
+R(nextEE(i)) = R(t_i0) + width(i,j_next(i)) * (PEnd(i,j_next(i)) - P(i,t_i0))
+```
+
+Define "next Estimated F" `nextEF(i) = R(nextEE(i))`.  These `nextEF`
+values do not change with the mere passage of time during which
+nothing changes.  They also do not change when `mu_fair(t)` and
+`mu(i,t)` change, as long as the equation `mu(i,t) = mu_fair(t)`
+continues to hold.  Thus, this representation is insulated from
+interference from other queues.
+
+Keep these queues in a data structure sorted by increasing `nextEF`.
+Enumerating the queues that have the minimal `nextEF` value can be
+done in `O(N_Min * log N)` compute time, where `N_Min` is the number
+of queues having that minimal `nextEF`.  Finding the one of those that
+is next in round-robin order can be done in `O(N_Min)` compute time.
+Given the high resolution of modern clocks, `N_Min` will very rarely
+be greater than 1.
+
+At time `t`, to translate a queue's `nextEF` value to an expected
+clock reading `EE` (for comparison with the expected `E` value from
+the easier category of queues), assume `mu_fair(t)` will hold steady
+into the future and reason as follows.
+
+```
+R(tau) - R(t) = (tau - t) * mu_fair(t)
+
+nextEF(i) = R(nextEE(i)) = R(t) + (nextEE(i) - t) * mu_fair(t)
+
+nextEE(i) = t + ( nextEF(i) - R(t) ) / mu_fair(t)
+```
+
+By maintaining these two sorted sets of queues, the runtime cost to
+find the next request to dispatch in the real world reduces from
+`O(N)` to `O(log N)` compute time.
+
+Thus, this implementation costs `O((1 + N_Delta) * log N)` compute
+time per request.
+
+
+### Fair Queuing for Server Requests, with max-min fairness and concurrent virtual execution
```

Comment:
I'm having hard time with following all the logic in this section.

So before we even agree to go this way, I would really like to better understand what kind of problems that we will have with "max-min fairness and serial execution" this would address.
I personally think that simpler implementation (even the current implementation is really complicated) is a huge advantage and it can reward some deficiencies if we understand them well.
I don't think I understand them well enough at this point to be able to say if this is really desired or not.

Can you please describe more concretely what are the exact problems with the "max-min fairness with serial execution" implementation and under what circumstance we may be facing them?